[Merged by Bors] - feat(Tactic/ExtractGoal + test/ExtractGoal): port `extract_goal` tactic

adomani opened the pull request:
This is a first PR porting some of the functionality of the `extract_goal` tactic from mathlib3.

---

It is my first attempt at porting a tactic, so do not restrain yourself from providing copious comments!

<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

alexjbest commented on Mathlib.lean:
```diff
@@ -2397,6 +2397,7 @@ import Mathlib.Tactic.Use
 import Mathlib.Tactic.WLOG
 import Mathlib.Tactic.Zify
 import Mathlib.Tactic.Zify.Attr
+import Mathlib.Tactic.extractGoal
```
I would expect the filename to still begin with a capital letter

adomani replied:
Changed, thanks!

alexjbest commented on Mathlib/Tactic/extractGoal.lean:
```diff
@@ -0,0 +1,80 @@
+import Std.Data.List.Basic
+
+/-!
+#  `extractGoal`: Format the current goal as a stand-alone example
```
The tactic syntax should probably be ` extract_goal` though

adomani replied:
Also, changed!

alexjbest commented on Mathlib/Tactic/extractGoal.lean:
````diff
@@ -0,0 +1,80 @@
+```
+
+* TODO: Deal with `let`
+* TODO: Add functionality to preduce a named `theorem` via `extractGoal thmName`
````
```suggestion
* TODO: Add functionality to produce a named `theorem` via `extractGoal thmName`
```

adomani replied:
Thanks!

alexjbest commented on Mathlib/Tactic/ExtractGoal.lean:
```diff
@@ -0,0 +1,80 @@
+-/
+
+/-- Translate a `BinderInfo` into the appropriate left-right pair of parentheses. -/
+def Lean.BinderInfo.toParens : BinderInfo → String × String
```
This already exists as `Lean.BinderInfo.brackets`

alexjbest commented:
I'm sorry I don't have time to make this comment more precise right now (so maybe its just a useless comment), but part of me can't help feeling some of this approach is a bit low-tech somehow? e.g. there is a lot of manual matching on and joining of strings. Although I don't know precisely how to do it immediately I would quite like to see a version of this PR that targets `Syntax` rather than strings/format, and uses the built in delaborators more somehow.
Of course if it gets good results as is then maybe we shouldn't worry too much right now, and maybe merge this and discuss whether a shorter/cleaner implementation is possible later, but if you were interested I think it would be good to play with a more Syntaxey version (I'd also like to try more seriously, but unfortunately can't this week)

adomani commented:
Alex, I see what you are saying, although I do not really know how to actually put it in practice.  I'm all for learning how to work with Syntax, but I am not really sure where to start.

When you have the time, I'd be happy to hear some suggestions!  In the meantime, I'll see that I can find!

kmill commented:
I'd prefer that we use delaborators and other constructions to build a declaration for this, rather than doing string manipulation. Here's an implementation that makes use of the signature pretty printer that's used by `#check`.

```lean
open Lean Meta Elab Tactic

/--
`extract_goal` formats the current goal as a stand-alone theorem or definition.
It tries to produce an output that can be copy-pasted and just work.

By default it cleans up the local context. To use the full local context, use `extract_goal*`.
-/
elab (name := extractGoal) "extract_goal" full?:&"*"? name?:(colGt ppSpace ident)? : tactic => do
  let name ← if let some name := name?
             then pure name.getId
             else mkAuxName ((← getCurrNamespace) ++ `extracted) 1
  let msg ← withoutModifyingEnv <| withoutModifyingState do
    let mut g ← getMainGoal
    unless full?.isSome do
      g ← g.cleanup
    (g, _) ← g.renameInaccessibleFVars
    (_, g) ← g.revert (clearAuxDeclsInsteadOfRevert := true) (← g.getDecl).lctx.getFVarIds
    let ty ← instantiateMVars (← g.getType)
    if ty.hasExprMVar then
      -- TODO: turn metavariables into new hypotheses?
      throwError "Extracted goal has metavariables: {ty}"
    let ty ← Term.levelMVarToParam ty
    let seenLevels := collectLevelParams {} ty
    let levels := (← Term.getLevelNames).filter
                    fun u => seenLevels.visitedLevel.contains (.param u)
    addAndCompile <| Declaration.axiomDecl
      { name := name
        levelParams := levels
        isUnsafe := false
        type := ty }
    let sig ← addMessageContext <| MessageData.ofPPFormat { pp := fun
                | some ctx => ctx.runMetaM <| PrettyPrinter.ppSignature name
                | none     => unreachable!
              }
    let cmd := if ← Meta.isProp ty then "theorem" else "def"
    pure m!"{cmd} {sig} := sorry"
  logInfo msg
```

adomani commented:
Kyle, thank you very much: your version is great!

I replaced mine with yours, added you as an author and mentioned some of the cases in which the extracted theorem is incorrect.

Note that my version gave incorrect results way more often!

adomani commented:
I will add a test that the extracted goal proves the given goal and emit a warning otherwise.  However, I do not have time to do it right now!

adomani commented:
I tried and failed.  I am unable to get my hands on the output of `logInfo msg`, in order to parse it again.  If I had that, then I would
* make sure that it parses with no issues, other than the `sorry`;
* check that the new declaration actually closes the initial goal.

I have added a couple of examples to the docs-module containing examples that would form good tests, if the above were implemented.

Still, I think that the tactic already works very well as is and these improvements could come later on.

An option would be to add something like `extract_goal [...] pp.all` to specify that you want the full, round-tripping expression.

kmill commented on Mathlib/Tactic/ExtractGoal.lean:
```diff
@@ -0,0 +1,128 @@
+
+/--
+`extract_goal` formats the current goal as a stand-alone theorem or definition.
+It tries to produce an output that can be copy-pasted and just work.
```
May as well put a warning that people shouldn't be *too* hopeful :wink: :
```suggestion
It tries to produce an output that can be copy-pasted and just work,
but its success depends on whether the expressions are amenable
to being unambiguously pretty printed.
```

I haven't tested it, but could you try it and if it works add this suggestion?
```
The tactic responds to pretty printing options.
For example, `set_option pp.all in extract_goal` gives the `pp.all` form.
```

Maybe we should add to the first line of the docstring too:
```
`extract_goal` formats the current goal as a stand-alone theorem or definition,
and `extract_goal name` uses the name `name` instead of an autogenerated one.
```

adomani replied:
The `set_option pp.all in extract_goal` trick is great!!

This works with no errors:
```lean
theorem extracted_1.{u_1} {α : Sort u_1} (a : α) : @Exists.{u_1} (α → α) fun (f : α → α) ↦ @Eq.{u_1} α (f a) a := sorry

example (a : α) : ∃ f : α → α, f a = a := by
  set_option pp.all true in
  extract_goal
  apply extracted_1

open Polynomial

theorem extracted_2 :
  @Eq.{1} (@Polynomial.{0} Nat Nat.semiring) (@Polynomial.X.{0} Nat Nat.semiring)
    (@Polynomial.X.{0} Nat Nat.semiring) := sorry


example : (X : Nat[X]) = X := by
  set_option pp.all true in
  extract_goal
  apply extracted_2
```

adomani replied:
I added the `pp.all` trick to the docs and a test.

adomani replied:
Btw, I checked and the polynomial example, failed also in Lean3.  However, the "exists" example actually inserted the type-ascription:
```lean
example {z : ℤ}: ∃ n : ℕ, z = n := begin
  extract_goal coes
end

-- produces:
lemma coes {z : ℤ} : ∃ (n : ℕ), z = ↑n :=
begin
  admit,
end
```
In either case, you could place `set_option pp.all true` before the *statement* and it would print it with that option.  It does not seem to be possible to use `pp.all` *inside* the proof, though.

kmill replied:
Exists doesn't show a type ascription in Lean 4 because it's implemented using an `app_unexpander` and is transforming a `fun`, but by default lambdas don't show types. I've been wanting to get around to fixing this for a while.

adomani commented:
@kmill, CI failed to build the tests, since `#guard_msgs` complained for the `pp.all` output.  Does CI remove `pp`-options and hence the test will check against the non-`pp.all` message?

I am running CI again to test this, but maybe you already know what the issue it!

adomani commented:
I looked closer at the diff: the current hypothesis is that what my computer displays as `fun ... ↦ ...`, CI displays as `fun ... => ...`.

adomani commented:
Yes, it seems to be this: on my local version, `#guard_msgs` complains that the outputs do not match, because of the `fun` arrow.  On GitHub, though, the PR has passed the tests, so that `#guard_msgs` liked what it received.

kmill commented:
Ok, that's a known issue then. Add `set_option pp.unicode.fun true` to the top of the tests file.
